### PR TITLE
Delegate data directory location in the sandbox to the controller

### LINKED ISSUE #747: fix defaultPrefix to honor a root ?
the [defaultPrefix](https://github.com/docker/libnetwork/blob/master/sandbox.go#L572) in libnetwork not honoring any specified root, when I have given root for graphdriver? but libnetwork still creates `/var/lib/docker` can we avoid this?

and this could be a problem when I ran different docker daemon with different driver / root flags simultaneously?

``` console
➸ sudo find /var/lib/docker/ -ls
 93174    0 drw-------   3 root     root           60 Nov  8 18:14 /var/lib/docker/
 93175    0 drw-------   3 root     root           60 Nov  8 18:14 /var/lib/docker/network
 93176    0 drw-------   2 root     root           60 Nov  8 21:41 /var/lib/docker/network/files
  1200    0 srw-------   1 root     root            0 Nov  8 21:41 /var/lib/docker/network/files/d10945c8a9706cc93c1b8074f20092d959a0b7d0a64b7cee1058bcdb91fd5c6e.sock
```

``` console
# part of docker daemon --help
  --exec-root=/var/run/docker          Root of the Docker execdriver
  -g, --graph=/var/lib/docker          Root of the Docker runtime
```

### PR DESCRIPTION
**Pull request by calavera**
Because this library should not assume that /var/lib/docker is
always the root path.

Fixes #747

Signed-off-by: David Calavera <david.calavera@gmail.com>

### CODE REVIEW

File: sandbox.go [hunk 1]
```diff
@@ -111,6 +111,17 @@ type containerConfig struct {
...
+	dataDir           string
+}
+
+// newContainerConfig initializes the configuration
+// using the defaultPrefix as the data directory.
+// That option can be overriden by using OptionDataDir
+// when a new sandbox is created.
+func newContainerConfig() containerConfig {
+	return containerConfig{
+		dataDir: defaultPrefix,
```
Comment (mrjana) [inline sandbox.go]: @calavera I don't think we need another dataDir config for every container. It's kind of daemon wide configuration and so it's appropriate to have it in `controller` and use it everywhere. In fact we already have this option as part of `controller`  here: https://github.com/docker/libnetwork/blob/master/config/config.go#L25. 

So we should make use of it and rework this PR based on that config?

**Comment by calavera** [comment]
@mrjana I totally agree with you. I redid the patch to use the controller's data directory and remove that default prefix.

**Comment by thaJeztah** [comment]
Looks like this may resolve https://github.com/docker/libnetwork/issues/747 ?

**Comment by mavenugo** [comment]
@calavera thanks for the fix. Could you also please take care of the hard-coding in `sandbox_externalkey_unix.go` ?

**Comment by calavera** [comment]
will do

**Comment by thaJeztah** [comment]
ping @calavera looks like you didn't fix `sandbox_externalkey_unix.go` yet :smile:

**Comment by calavera** [comment]
@thaJeztah I did, but changing that requires more work than I expected, and I'm slightly busy with the api right now :smiley_cat:

**Comment by thaJeztah** [comment]
@calavera yup, I can imagine. I just landed here through https://github.com/docker/docker/issues/18508, and noticed it wasn't merged/updated yet :+1:

**Comment by calavera** [comment]
@mavenugo @thaJeztah I'm not sure we can remove that hardcoded path. It's used to store the sockets for reexecution. If we remove it, we'd need to send the full path to the socket as the third argument to set the network key. I really have no idea where we run that reexecution.

**Comment by mavenugo** [comment]
@calavera I will take a look at this as well. Since you have already completed the work on other hard-coded paths, I think we can get this one in. I will handle the sandbox_externalkey_unix.go.

**Comment by mavenugo** [comment]
@thaJeztah @calavera I dont think this is critical for 1.10 release. Given the intrusive changes and potentially causing migration issues (if we start using another location to look for local states), I would keep this PR out of 1.10 release.

**Comment by thaJeztah** [comment]
@mavenugo wondering if this can cause corruption of the local-kv.db if somebody either switches graph-directory, or has two daemons on the same host, but different graph-dir?

**Comment by mrjana** [comment]
@thaJeztah How do you mean when you say "if somebody either switches graph-directory"? If somebody switches graph-directory across docker restarts we are talking about different local-kv.db instances. If two docker hosts run in different graph-dir again this is a supported configuration and we full support it. In fact we create local-kv.db in exactly the graph dir option used in daemon option. 

Actually there is no hard coding at all. The only thing that is hard coded is the location of the unix domain socket that libnetwork listens to for pre/post start hooks. But that unix domain socket is named using random controller id which cannot collide. 

So while this is a good PR to merge in the master, the only thing that is hard coded is the unix domain socket path. Hence I don't think this is a candidate for 1.10 since the change is more pervasive and reward is less than the risk.

**Comment by mavenugo** [comment]
@mrjana its my mistake that I incorrectly pointed about migration, while that is not an issue in this case.
@thaJeztah I agree with @mrjana we took care of the local-kv.db stored in the correct root folder (pointed by -g) in 1.9 release.

**Comment by thaJeztah** [comment]
@mrjana apologies, I think I misinterpreted this part https://github.com/docker/libnetwork/blob/317d527fd6628f51662e1c1c1c0ac9f1b911d6c5/datastore/datastore.go#L118-L137 also initializing the database in that location

**Comment by mrjana** [comment]
@thaJeztah Yeah that's the default default. But a non-empty data directory argument from engine will always override it as can be seen here:
https://github.com/docker/libnetwork/blob/317d527fd6628f51662e1c1c1c0ac9f1b911d6c5/datastore/datastore.go#L159

**Comment by calavera** [comment]
@mavenugo how hard would it be to change the third parameter in the reexec function to take the whole path to the socket? that seems like the easiest solution, but I still don't know the implications of changing one of those arguments?

**Comment by mavenugo** [comment]
@calavera I think that is a reasonable and safe change. To give you a history of this change, it is supposed to be called from the hook callback from runc (when it runs as an external process). It is currently embedded as the libcontainer library in docker and hence the reexec path is not exercised by Docker. But when it does, it can just call the re-exec with the said parameters so that libnetwork can continue doing its sandbox configuration phase. Since runc library is in the same process context as the docker engine, we call the `SetKey` directly from docker and hence bypassing the entire reexec path.
Libnetwork is designed to independent of how docker engine makes use of runc.
With that history, I think it is safe to do that change and I would have done the same if I handled it.

@calavera @thaJeztah do you think this is a mandatory PR for 1.10 ? 
We can absolutely review and merge it (after fixing the sandbox_externalkey) in master. We have to determine the importance of it for the 1.10 branch.

**Comment by calavera** [comment]
okay, that makes more sense. I'm going to make the smallest changes I can to support the path as the third argument.

I think we should try to have this in 1.10. Give me a few minutes to put the patch together.

**Comment by tiborvass** [comment]
@mavenugo I'm just here to give my vote for fixing this for 1.10, especially if it's doable without much risk.

**Comment by mrjana** [comment]
I don't agree that this should be part of 1.10. If it was merged before code freeze may be but not now.

**Comment by tiborvass** [comment]
@mrjana this is a bug, not a docker api change.

**Comment by mrjana** [comment]
@tiborvass This is code cleanup with no user visible bug.

**Comment by mavenugo** [comment]
@mrjana it is an user visible bug : https://github.com/docker/libnetwork/issues/747

**Comment by mrjana** [comment]
I don't attribute #747 as a user visible bug as we don't even document anywhere where we are going to place that UDS sock. But all of this is just one man's opinion.

**Comment by tiborvass** [comment]
@mrjana docker/docker#18508

**Comment by mavenugo** [comment]
@mrjana it is not just the UDS sock. IT is also `resolvConfPath` and `hostsPath`

**Comment by mrjana** [comment]
@mavenugo It's not. It's just UDS sock. We always get resolvConfPath and hostsPath from engine

**Comment by calavera** [comment]
why do we even start the reexec process if we're not currently using it? can we disable that?

**Comment by mavenugo** [comment]
@calavera we did that because libnetwork was originally designed to be used independent of the decision to integrate with runc as an external process. If there are no such plans, I think it is safe to even turn off the functionality for now.

**Comment by mrjana** [comment]
ping @calavera. Can we get this blessed by CI?

**Comment by luxas** [comment]
What happened to this PR @calavera and @mavenugo? 
It would be good to have merged as it's causing issues for Kubernetes deployments with two daemons.

**Comment by mavenugo** [comment]
@luxas this issue shouldnt cause any issue with 2 daemons with 2 different `-g` locations. The only issue here is the fact that, the sock is stored in the static location. but that shouldn't cause any specific daemon issue.

> it's causing issues for Kubernetes deployments with two daemons.

Can you please explain the issue ?

**Comment by luxas** [comment]
@mavenugo Please see: https://github.com/docker/docker/issues/21156

**Pull request closed without merging**